sympy/core/mul.py

```diff
@@ -1058,11 +1058,31 @@ def _eval_is_real(self):
             return real  # doesn't matter what zero is
 
     def _eval_is_imaginary(self):
-        z = self.is_zero
-        if z:
-            return False
-        elif z is False:
-            return (S.ImaginaryUnit*self).is_real
+        real = True
+        zero = one_neither = False
+
+        for t in self.args:
+            if t.is_imaginary:
+                real = not real
+            elif t.is_real:
+                if zero is False:
+                    zero = fuzzy_not(t.is_nonzero)
+                    if zero:
+                        return False
+            elif t.is_real is False:
+                if one_neither:
+                    return  # complex terms might cancel
+                one_neither = True
+            else:
+                return
+
+        if zero is False:
+            if one_neither:  # N = a+I*b or I*b
+                if real:
+                    return  # r*N is like N: could be either
+                return False  # neither I*N values is imaginary
+            if not real:
+                return True
```

Comment:
This looks like code that you copied from somewhere else.
If that is the case, it would be a good idea to put the code into a function and call it from the original place and from here. Otherwise we'll end with lots of duplicated code in our codebase.